### PR TITLE
Use reentrant mode when invoking prims, delete global prim_fake_mode

### PR DESCRIPTION
**Pull request by eellison**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84310
* #84096
* __->__ #84090

Maybe I should be using the meta_impl instead of the prim_impl, but it's not terribly clear why, since the prim impl will be better tested and should work under the re-entrant FakeTensorMode.

Fixes https://github.com/pytorch/pytorch/issues/78613 in the process

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

File: torch/_prims/__init__.py [hunk 1]
```diff
@@ -259,22 +238,13 @@ def TensorMeta(
...
-    else:
-        mode = get_prim_fake_mode()
-
-    if device.type == "meta":
-        return torch.empty_strided(shape, strides, dtype=dtype, device="meta")
+    # SymInt doesnt support empty_strided yet
+    if any(
+        isinstance(inp, torch.SymIntNode) for inp in itertools.chain(shape, strides)
+    ):
+        return torch.empty(shape, dtype, device=device)
```
Comment (bdhirsh) [inline torch/_prims/__init__.py:245]: Should this just unconditionally call `empty_strided` once it has symint support?

I can try to update it once https://github.com/pytorch/pytorch/pull/81238/files lands (I'm updating it today)

File: test/test_ops.py [hunk 2]
```diff
@@ -167,7 +167,9 @@ def test_numpy_ref(self, device, dtype, op):
     @onlyNativeDeviceTypes
     @ops(python_ref_db)
     def test_python_ref_meta(self, device, dtype, op):
-        mode = torch._prims.get_prim_fake_mode()
+        mode = FakeTensorMode()
+        with mode:
```
Comment (samdow) [inline test/test_ops.py:171]: nit: slight preference for `with FakeTensorMode() as mode:` here and L394

**Review by samdow — APPROVED** [review]

**Comment by eellison** [comment]
I'm running into an issue with:
`prims.cos` registers a decomposition for `aten.cos` to call into `prims.cos`. `prims.cos` internally invokes `aten.cos`... So you have an infinite loop. I did a somewhat temporary change here just to try to get things to pass that will invoke the meta kernel for `prims.cos` instead of the aten kernel if we're in the middle of running `prims.cos`. Open to suggestions on other changes here. 

I'm running into an error with the SymInt tests. 
```
> torch.empty(shape, dtype=dtype, device=device)
RuntimeError: SymIntArrayRef expected to contain only concrete integers
```

File: torch/_prims/__init__.py [hunk 3]
```diff
@@ -259,22 +238,13 @@ def TensorMeta(
...
-    else:
-        mode = get_prim_fake_mode()
-
-    if device.type == "meta":
-        return torch.empty_strided(shape, strides, dtype=dtype, device="meta")
+    # SymInt doesnt support empty_strided yet
+    if any(
+        isinstance(inp, torch.SymIntNode) for inp in itertools.chain(shape, strides)
+    ):
+        return torch.empty(shape, dtype=dtype, device=device)
```
Comment (eellison) [inline torch/_prims/__init__.py:245]: hmm, this is failing now..

Comment (eellison) [reply torch/_prims/__init__.py:245]: when I invoke this with `torch.empty([2, 3], dtype=dtype, device=device)` here a FakeTensor is returned successfully but `torch.ops.aten.empty.SymInt` with SymInt inputs fails with `RuntimeError: SymIntArrayRef expected to contain only concrete integers

cc @horace, @Krovatkin

**Comment by eellison** [comment]
cc @malfet  @seemethere this is failing with 
```
ios-12-5-1-arm64-coreml - Unauthorized
ios-12-5-1-arm64-custom-ops - Unauthorized
[macos-12-py3-x86-64-build](https://circleci.com/gh/pytorch/pytorch/17066636?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
[macos-12-py3-x86-64-test-2-2-default](https://circleci.com/gh/pytorch/pytorch/17066711?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
[macos-12-py3-arm64-build](https://circleci.com/gh/pytorch/pytorch/17066635?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) - Success
ios-12-5-1-arm64-metal - Unauthorized
ios-12-5-1-arm64 - Unauthorized
``` 

infra error ?

**Comment by malfet** [comment]
@eellison yes, please ignore

**Pull request closed without merging**